haigha/connection.py

```diff
@@ -405,9 +409,18 @@ def read_frames(self):
         self._channels[0].send_heartbeat()
 
         data = self._transport.read(self._heartbeat)
+        current_time = time.time()
+
         if data is None:
+            # Wait for 2 heartbeat intervals before giving up. See AMQP 4.2.7:
+            # "If a peer detects no incoming traffic (i.e. received octets) for two heartbeat intervals or longer,
+            # it should close the connection"
+            if self._heartbeat and self._last_octet_time and (current_time-self._last_octet_time > 2*self._heartbeat):
+                msg = 'Heartbeats not received from %s for %d seconds' % (self._host, 2*self._heartbeat)
+                self.transport_closed(msg=msg)
+                raise ConnectionClosed('Connection is closed: ' + msg)
             return
-
+        self._last_octet_time = time.time()
```

Comment:
Might as well use `current_time` since we've already called out to the kernel to fetch it and this block is executed frequently.